src/engine/QueryPlanner.cpp

```diff
@@ -2029,18 +2042,63 @@ auto QueryPlanner::createSpatialJoin(
   return plan;
 }
 
+// _____________________________________________________________________________________________________________________
+auto QueryPlanner::createJoinWithUnitedTransitivePath(
+    SubtreePlan a, SubtreePlan b,
+    const std::vector<std::array<ColumnIndex, 2>>& jcs) const
+    -> std::vector<SubtreePlan> {
+  auto aUnionPath =
+      std::dynamic_pointer_cast<Union>(a._qet->getRootOperation());
+  auto bUnionPath =
+      std::dynamic_pointer_cast<Union>(b._qet->getRootOperation());
+  std::vector<SubtreePlan> candidates{};
+  auto findCandidates = [this, &candidates, &jcs](Union& unionOperation,
+                                                  const SubtreePlan& other,
+                                                  bool unionIsLeft) {
+    auto* qec = other._qet->getRootOperation()->getExecutionContext();
+
+    // Clone `other` once, to avoid using it twice in the same query plan.
+    SubtreePlan clonedOther = other;
+    clonedOther._qet = other._qet->clone();
+    SubtreePlan tmpPlan = other;
+    const auto& leftCandidate = unionIsLeft ? tmpPlan : other;
+    const auto& rightCandidate = unionIsLeft ? clonedOther : tmpPlan;
+
+    tmpPlan._qet = unionOperation.leftChild();
+    auto joinedLeft = createJoinCandidates(leftCandidate, rightCandidate, jcs);
+    tmpPlan._qet = unionOperation.rightChild();
+    auto joinedRight = createJoinCandidates(leftCandidate, rightCandidate, jcs);
+
+    for (const auto& leftPlan : joinedLeft) {
+      for (const auto& rightPlan : joinedRight) {
+        SubtreePlan candidate =
+            makeSubtreePlan<Union>(qec, leftPlan._qet, rightPlan._qet);
+        mergeSubtreePlanIds(candidate, leftPlan, rightPlan);
+        candidates.push_back(std::move(candidate));
+      }
+    }
+  };
+  if (aUnionPath) {
+    findCandidates(*aUnionPath, b, true);
+  }
+  if (bUnionPath) {
+    findCandidates(*bUnionPath, a, false);
+  }
+  return candidates;
+}
+
 // __________________________________________________________________________________________________________________
 auto QueryPlanner::createJoinWithTransitivePath(
     SubtreePlan a, SubtreePlan b,
-    const std::vector<std::array<ColumnIndex, 2>>& jcs)
-    -> std::optional<SubtreePlan> {
+    const std::vector<std::array<ColumnIndex, 2>>& jcs) const
+    -> std::vector<SubtreePlan> {
   auto aTransPath = std::dynamic_pointer_cast<const TransitivePathBase>(
       a._qet->getRootOperation());
   auto bTransPath = std::dynamic_pointer_cast<const TransitivePathBase>(
       b._qet->getRootOperation());
 
   if (!(aTransPath || bTransPath)) {
-    return std::nullopt;
+    return createJoinWithUnitedTransitivePath(std::move(a), std::move(b), jcs);
```

Comment:
This currently also unites all unions With Everything, 
as there is not a single check for actual transitive paths somewhere.
This is definitely not something we want to have here, and probably we also need a more complex mechanism for really good plans in this mechanism in general.
So for now:
Only if the child of the UNION is an unbound transitive path you should apply this optimization.